src/core/read/preframe.cc

```diff
@@ -0,0 +1,227 @@
+  return columns_.end();
+}
+
+PreFrame::const_iterator PreFrame::begin() const {
+  return columns_.begin();
+}
+
+PreFrame::const_iterator PreFrame::end() const {
+  return columns_.end();
+}
+
+
+
+
+//------------------------------------------------------------------------------
+// Column parse types
+//------------------------------------------------------------------------------
+
+std::vector<PT> PreFrame::get_ptypes() const {
+  std::vector<PT> res(columns_.size(), PT::Mu);
+  save_ptypes(res);
+  return res;
+}
+
+void PreFrame::save_ptypes(std::vector<PT>& types) const {
+  xassert(types.size() == columns_.size());
+  size_t i = 0;
+  for (auto& col : columns_) {
+    types[i++] = col.get_ptype();
+  }
+}
+
+bool PreFrame::are_same_ptypes(std::vector<PT>& types) const {
+  xassert(types.size() == columns_.size());
+  size_t i = 0;
+  for (auto& col : columns_) {
+    if (types[i++] != col.get_ptype()) return false;
+  }
+  return true;
+}
+
+void PreFrame::set_ptypes(const std::vector<PT>& types) {
+  xassert(types.size() == columns_.size());
+  size_t i = 0;
+  for (auto& col : columns_) {
+    col.force_ptype(types[i++]);
+  }
+}
+
+void PreFrame::reset_ptypes() {
+  for (auto& col : columns_) {
+    col.force_ptype(PT::Mu);
+  }
+}
+
+const char* PreFrame::print_ptypes() const {
+  const ParserInfo* parsers = ParserLibrary::get_parser_infos();
+  static const size_t N = 100;
+  static char out[N + 1];
+  char* ch = out;
+  size_t ncols = columns_.size();
+  size_t tcols = ncols <= N? ncols : N - 20;
+  for (size_t i = 0; i < tcols; ++i) {
+    *ch++ = parsers[columns_[i].get_ptype()].code;
+  }
+  if (tcols != ncols) {
+    *ch++ = ' ';
+    *ch++ = '.';
+    *ch++ = '.';
+    *ch++ = '.';
+    *ch++ = ' ';
+    for (size_t i = ncols - 15; i < ncols; ++i)
+      *ch++ = parsers[columns_[i].get_ptype()].code;
+  }
+  *ch = '\0';
+  return out;
+}
+
+
+
+
+//------------------------------------------------------------------------------
+// Aggregate column stats
+//------------------------------------------------------------------------------
+
+size_t PreFrame::n_columns_in_output() const {
+  size_t n = 0;
+  for (const auto& col : columns_) {
+    n += col.is_in_output();
+  }
+  return n;
+}
+
+size_t PreFrame::n_columns_in_buffer() const {
+  size_t n = 0;
+  for (const auto& col : columns_) {
+    n += col.is_in_buffer();
+  }
+  return n;
+}
+
+size_t PreFrame::n_columns_to_reread() const {
+  size_t n = 0;
+  for (const auto& col : columns_) {
+    n += col.is_type_bumped();
+  }
+  return n;
+}
+
+
+size_t PreFrame::total_allocsize() const {
+  size_t allocsize = sizeof(*this);
+  for (const auto& col : columns_) {
+    allocsize += col.memory_footprint();
+  }
+  return allocsize;
+}
+
+
+
+
+//------------------------------------------------------------------------------
+// Finalizing
+//------------------------------------------------------------------------------
+using dtptr = std::unique_ptr<DataTable>;
```

Comment:
Since we're now doing some code cleaning, what about moving this definition to `_dt.h` (or some other header file), so that we include this file from everywhere else? Right now I see at least four duplicated definitions of `dtptr`, namely, `reader.h`, `aggregator.h`, `dt_ftrl_base.h` and `preframe.cc`.

Reply:
yep, if you see duplicated definitions like that, then it's perfectly fine to add them to the _dt.h file. This file is just a collection of all forward declarations (all that I've found so far). It can be expanded in future revisions.